File: internal/handlers/company_handler.go
package handlers

import (
	"context"
	"net/http"
	"time"

	"chat-backend/internal/database"
	"chat-backend/internal/models"

	"github.com/gin-gonic/gin"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
	"go.mongodb.org/mongo-driver/mongo/options"
)

type CompanyHandler struct {
	db *database.Database
}

func NewCompanyHandler(db *database.Database) *CompanyHandler {
	return &CompanyHandler{db: db}
}

// isValidCategory checks whether the given category is in the allowed list
func isValidCategory(cat string) bool {
	for _, c := range models.CompanyCategories {
		if c == cat {
			return true
		}
	}
	return false
}

// CreateCompany — POST /companies
func (h *CompanyHandler) CreateCompany(c *gin.Context) {
	userID, _ := c.Get("user_id")
	userIDObj := userID.(primitive.ObjectID)

	var req struct {
		Name        string `json:"name" binding:"required"`
		Category    string `json:"category" binding:"required"`
		Description string `json:"description"`
		Website     string `json:"website"`
	}
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	if !isValidCategory(req.Category) {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid category"})
		return
	}

	now := time.Now()
	company := models.Company{
		ID:          primitive.NewObjectID(),
		UserID:      userIDObj,
		Name:        req.Name,
		Category:    req.Category,
		Description: req.Description,
		Website:     req.Website,
		CreatedAt:   now,
		UpdatedAt:   now,
	}

	if _, err := h.db.MongoDB.Collection("companies").InsertOne(context.Background(), company); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to create company"})
		return
	}

	c.JSON(http.StatusCreated, company)
}

// GetMyCompanies — GET /companies/me
func (h *CompanyHandler) GetMyCompanies(c *gin.Context) {
	userID, _ := c.Get("user_id")
	userIDObj := userID.(primitive.ObjectID)

	opts := options.Find().SetSort(bson.D{{Key: "created_at", Value: -1}})
	cursor, err := h.db.MongoDB.Collection("companies").Find(
		context.Background(),
		bson.M{"user_id": userIDObj},
		opts,
	)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch companies"})
		return
	}
	defer cursor.Close(context.Background())

	var companies []models.Company
	if err := cursor.All(context.Background(), &companies); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to decode companies"})
		return
	}
	if companies == nil {
		companies = []models.Company{}
	}

	c.JSON(http.StatusOK, companies)
}

// GetUserCompanies — GET /companies/user/:user_id  (public — any authenticated user can view)
func (h *CompanyHandler) GetUserCompanies(c *gin.Context) {
	userIDStr := c.Param("user_id")
	userIDObj, err := primitive.ObjectIDFromHex(userIDStr)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid user ID"})
		return
	}

	opts := options.Find().SetSort(bson.D{{Key: "created_at", Value: -1}})
	cursor, err := h.db.MongoDB.Collection("companies").Find(
		context.Background(),
		bson.M{"user_id": userIDObj},
		opts,
	)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch companies"})
		return
	}
	defer cursor.Close(context.Background())

	var companies []models.Company
	if err := cursor.All(context.Background(), &companies); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to decode companies"})
		return
	}
	if companies == nil {
		companies = []models.Company{}
	}

	c.JSON(http.StatusOK, companies)
}

// UpdateCompany — PUT /companies/:company_id
func (h *CompanyHandler) UpdateCompany(c *gin.Context) {
	userID, _ := c.Get("user_id")
	userIDObj := userID.(primitive.ObjectID)

	companyIDStr := c.Param("company_id")
	companyIDObj, err := primitive.ObjectIDFromHex(companyIDStr)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid company ID"})
		return
	}

	var req struct {
		Name        string `json:"name"`
		Category    string `json:"category"`
		Description string `json:"description"`
		Website     string `json:"website"`
	}
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	if req.Category != "" && !isValidCategory(req.Category) {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid category"})
		return
	}

	// Only owner can update
	existing := models.Company{}
	if err := h.db.MongoDB.Collection("companies").FindOne(
		context.Background(),
		bson.M{"_id": companyIDObj},
	).Decode(&existing); err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Company not found"})
		return
	}
	if existing.UserID != userIDObj {
		c.JSON(http.StatusForbidden, gin.H{"error": "Not allowed"})
		return
	}

	update := bson.M{"updated_at": time.Now()}
	if req.Name != "" {
		update["name"] = req.Name
	}
	if req.Category != "" {
		update["category"] = req.Category
	}
	update["description"] = req.Description
	update["website"] = req.Website

	if _, err := h.db.MongoDB.Collection("companies").UpdateOne(
		context.Background(),
		bson.M{"_id": companyIDObj},
		bson.M{"$set": update},
	); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to update company"})
		return
	}

	// Return updated document
	updated := models.Company{}
	h.db.MongoDB.Collection("companies").FindOne(context.Background(), bson.M{"_id": companyIDObj}).Decode(&updated)
	c.JSON(http.StatusOK, updated)
}

// DeleteCompany — DELETE /companies/:company_id
func (h *CompanyHandler) DeleteCompany(c *gin.Context) {
	userID, _ := c.Get("user_id")
	userIDObj := userID.(primitive.ObjectID)

	companyIDStr := c.Param("company_id")
	companyIDObj, err := primitive.ObjectIDFromHex(companyIDStr)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid company ID"})
		return
	}

	// Only owner can delete
	existing := models.Company{}
	if err := h.db.MongoDB.Collection("companies").FindOne(
		context.Background(),
		bson.M{"_id": companyIDObj},
	).Decode(&existing); err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Company not found"})
		return
	}
	if existing.UserID != userIDObj {
		c.JSON(http.StatusForbidden, gin.H{"error": "Not allowed"})
		return
	}

	if _, err := h.db.MongoDB.Collection("companies").DeleteOne(
		context.Background(),
		bson.M{"_id": companyIDObj},
	); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to delete company"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Company deleted"})
}

// GetCategories — GET /companies/categories  (public — returns allowed category list)
func (h *CompanyHandler) GetCategories(c *gin.Context) {
	c.JSON(http.StatusOK, gin.H{"categories": models.CompanyCategories})
}

File: src/handlers/companyHandler.js
'use strict';

const { ObjectId } = require('mongodb');
const { getDB } = require('../database');

// ─── Constants ────────────────────────────────────────────────────────────────

const CompanyCategories = [
  'Technology',
  'Food & Beverage',
  'Retail',
  'Healthcare',
  'Education',
  'Finance',
  'Real Estate',
  'Transportation',
  'Entertainment',
  'Manufacturing',
  'Construction',
  'Agriculture',
  'Energy',
  'Telecommunications',
  'Tourism',
  'Media',
  'Fashion',
  'Sports',
  'Legal',
  'Other',
];

// ─── Handlers ─────────────────────────────────────────────────────────────────

async function createCompany(req, res) {
  try {
    const db = getDB();
    const userId = req.userId;
    const { name, category, description, website } = req.body || {};

    if (!name) return res.status(400).json({ error: 'name is required' });
    if (!category) return res.status(400).json({ error: 'category is required' });
    if (!CompanyCategories.includes(category)) {
      return res.status(400).json({ error: `Invalid category. Must be one of: ${CompanyCategories.join(', ')}` });
    }

    const now = new Date();
    const companyDoc = {
      user_id: userId,
      name,
      category,
      description: description || '',
      website: website || null,
      created_at: now,
      updated_at: now,
    };

    const insertResult = await db.collection('companies').insertOne(companyDoc);
    companyDoc._id = insertResult.insertedId;

    return res.status(201).json({ ...companyDoc, id: insertResult.insertedId.toString() });
  } catch (err) {
    console.error('createCompany error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function getMyCompanies(req, res) {
  try {
    const db = getDB();
    const userId = req.userId;

    const companies = await db.collection('companies').find({ user_id: userId })
      .sort({ created_at: -1 })
      .toArray();

    return res.json(companies.map(c => ({ ...c, id: c._id.toString() })));
  } catch (err) {
    console.error('getMyCompanies error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function getUserCompanies(req, res) {
  try {
    const db = getDB();
    const targetUserIdStr = req.params.user_id;

    const companies = await db.collection('companies').find({ user_id: targetUserIdStr })
      .sort({ created_at: -1 })
      .toArray();

    return res.json(companies.map(c => ({ ...c, id: c._id.toString() })));
  } catch (err) {
    console.error('getUserCompanies error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function updateCompany(req, res) {
  try {
    const db = getDB();
    const userId = req.userId;
    const companyIdStr = req.params.company_id;
    const { name, category, description, website } = req.body || {};

    let companyObjId;
    try {
      companyObjId = new ObjectId(companyIdStr);
    } catch (_) {
      return res.status(400).json({ error: 'Invalid company ID' });
    }

    const company = await db.collection('companies').findOne({ _id: companyObjId });
    if (!company) return res.status(404).json({ error: 'Company not found' });

    if (company.user_id !== userId) {
      return res.status(403).json({ error: 'Forbidden: not the owner' });
    }

    if (category && !CompanyCategories.includes(category)) {
      return res.status(400).json({ error: `Invalid category. Must be one of: ${CompanyCategories.join(', ')}` });
    }

    const setFields = { updated_at: new Date() };
    if (name !== undefined) setFields.name = name;
    if (category !== undefined) setFields.category = category;
    if (description !== undefined) setFields.description = description;
    if (website !== undefined) setFields.website = website;

    await db.collection('companies').updateOne(
      { _id: companyObjId },
      { $set: setFields }
    );

    const updated = await db.collection('companies').findOne({ _id: companyObjId });
    return res.json({ ...updated, id: updated._id.toString() });
  } catch (err) {
    console.error('updateCompany error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function deleteCompany(req, res) {
  try {
    const db = getDB();
    const userId = req.userId;
    const companyIdStr = req.params.company_id;

    let companyObjId;
    try {
      companyObjId = new ObjectId(companyIdStr);
    } catch (_) {
      return res.status(400).json({ error: 'Invalid company ID' });
    }

    const company = await db.collection('companies').findOne({ _id: companyObjId });
    if (!company) return res.status(404).json({ error: 'Company not found' });

    if (company.user_id !== userId) {
      return res.status(403).json({ error: 'Forbidden: not the owner' });
    }

    await db.collection('companies').deleteOne({ _id: companyObjId });

    return res.json({ message: 'Company deleted' });
  } catch (err) {
    console.error('deleteCompany error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function getCategories(req, res) {
  return res.json({ categories: CompanyCategories });
}

module.exports = {
  createCompany,
  getMyCompanies,
  getUserCompanies,
  updateCompany,
  deleteCompany,
  getCategories,
  CompanyCategories,
};
